### PR TITLE
Enable ocaml code formatting with `dune fmt`

### PR DESCRIPTION
**Pull request by yfyf**
This is a preview of automatic code formatting using `ocamlformat` / `dune fmt`

I tried to tweak it (see `.ocamlformat`), but a couple of issues I see with all 3 `ocamlformat` profiles (and that I am not sure are configurable):
- short record value definitions (`let a = { ... }`) get squashed into a single line, which is not good for readability (therefore currently `margin = 80`, to prevent this as much as possible)
- similar situation with pipe operators (`|>`)
- `dune` files are formatted weirdly too (sometimes it breaks libraries into separate lines, sometimes it doesn't?), did not try to tweak that

In general, `ocamlformat` is meant to produce a canonical format regardless of how the input source is formatted, so it does not preserve any manual breaks/whitepsace. 

@knuton @guyonvarch please check it out and see what you think.

You can experiment with the options here: https://gpetiot.github.io/ocamlformat-preview/

### CODE REVIEW

File: controller/bindings/connman/connman.ml [hunk 1]
```diff
@@ -445,123 +462,156 @@ struct
...
-      pure (fun name type' state strength favorite autoconnect ipv4 ipv4_user_config ipv6 ethernet proxy nameservers security ->
-          { _proxy = OBus_proxy.make ~peer:(OBus_context.sender context) ~path:path
-          ; _manager = manager
-          ; id = path |> CCList.last 1 |> CCList.hd
-          ; name ; type'; state; strength; favorite; autoconnect
-          ; ipv4 = (if Option.is_some ipv4_user_config then ipv4_user_config else ipv4)
-          ; ipv6; ethernet; proxy; nameservers; security
-          })
+      pure
+        (fun
```
Comment (yfyf) [inline controller/bindings/connman/connman.ml:490]: wat

File: controller/tests/server/update/update_tests.ml [hunk 2]
```diff
@@ -196,33 +192,44 @@ let () =
            [
              (* BOOTED = PRIMARY in all these *)
              Alcotest_lwt.test_case "Both slots out of date -> Update" `Quick
-               (fun _ () -> Scenario.run both_out_of_date);
+               (fun _ () -> Scenario.run both_out_of_date
+             );
```
Comment (yfyf) [inline controller/tests/server/update/update_tests.ml:196]: This looks like a `ocamlformat` bug?

File: controller/bindings/connman/connman.ml [hunk 3]
```diff
@@ -96,70 +85,83 @@ struct
...
-            in
-            let%lwt () = Logs_lwt.err ~src:log_src (fun m -> m
-               "Manager is requesting additional missing input(s): %s"
-               expected_properties_str
-            )
-            in
-            return @@ Error (MissingRequestedInputs mandatory_inputs)
+    | Passphrase p, [ "Passphrase" ] ->
+        return
+        @@ Ok [ ("Passphrase", p |> OBus_value.C.(make_single basic_string)) ]
```
[the remark this answers is not in the document]
Comment (knuton) [reply controller/bindings/connman/connman.ml:133]: Me too

**Comment by yfyf** [comment]
> I prefer lists with the `ocamlformat` profile

I see the visual appeal, but I find that it is more cumbersome to edit, because with `conventional` swapping list/collection members is simply swapping _lines_. With `ocamlformat` you cannot swap the first line with others, because the opening bracket `[` gets moved. But it's a minor thing, if you feel strongly about it, I wouldn't mind `ocamlformat`.

File: controller/bindings/curl/curl.ml [hunk 4]
```diff
@@ -13,30 +13,16 @@ let pretty_print_error error =
...
-
-  | ChannelClosed err ->
-      Printf.sprintf "channel closed: %s" err
-
-  | Exception err ->
-      Printf.sprintf "exception: %s" err
+      Printf.sprintf "curl error: %s (non-zero exit code %d)" (String.trim err)
+        n
+  | ProcessKill n -> Printf.sprintf "curl killed by signal %d" n
+  | ProcessStop n -> Printf.sprintf "curl stopped by signal %d" n
```
[the remark this answers is not in the document]
Comment (yfyf) [reply controller/bindings/curl/curl.ml:21]: Are you sure? I tried various combos of `break-cases` and they resulted in inconsistent breaking.

**Comment by yfyf** [comment]
> Use space around elements?

Don't have a strong opinion about it, but would opt for less customization of the default profile settings to avoid causing unexpected formatting.

File: controller/bindings/curl/curl.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (yfyf) [reply controller/bindings/curl/curl.ml:21]: Oh, right, but then it breaks even in cases when all the cases fit within a single line. That's not great too, I think.

E.g. it makes this thing becomes much less readable:
https://github.com/dividat/playos/commit/75aa8c8a2ccee2a4a4240317eaf1f8f82e38079f#diff-70886782e5ab141f7eb9c6f4df99f6154decb03ac9848233b3b9b48c95337e68R456

[the remark this answers is not in the document]
Comment (yfyf) [reply controller/bindings/curl/curl.ml]: That makes sense. I cherry-picked your commit 75aa8c8a2ccee2a4a4240317eaf1f8f82e38079f here.

File: controller/bindings/connman/connman.ml (hunk 1 above)
Comment (yfyf) [reply controller/bindings/connman/connman.ml:511]: I guess the issue here is not the formatting, but the `fun` with 10+ arguments.

**Comment by yfyf** [comment]
@guyonvarch anything else that sticks out / looks annoying? The current config is decent enough for me and I think consistency is more important than a specific style.

@knuton we have converged towards something, please take another look.

**Comment by yfyf** [comment]
> Using `field-space = loose` in [5b92642](https://github.com/dividat/playos/commit/5b92642248d7eaca1b0bff3d822d2beb09d11f97), this is looking good!

Cherry-picked

File: controller/bindings/connman/connman.mli [hunk 5]
```diff
@@ -38,21 +38,20 @@ end
...
   (** Input that the agent may provide to connect to a network.
 
       Note that not all possible inputs are supported and thus connecting to some networks is not possible (e.g. WPS). See the ConnMan Agent API documentation for more information.
   *)
   type input =
-    | None (** No input *)
-    | Passphrase of string (** The passphrase for authentication. For example a WEP key, a PSK passphrase or a passphrase for EAP authentication methods.*)
+    | None  (** No input *)
+    | Passphrase of string
+        (** The passphrase for authentication. For example a WEP key, a PSK passphrase or a passphrase for EAP authentication methods.*)
```
Comment (knuton) [inline controller/bindings/connman/connman.mli:48]: Maybe putting comments above would yield less weird formatting?

File: controller/bindings/zerotier/zerotier.ml [hunk 6]
```diff
@@ -2,38 +2,25 @@ open Lwt
...
       with
       | RequestSuccess (_, body) ->
           let open Ezjsonm in
-          from_string body
-          |> get_dict
-          |> List.assoc "address"
-          |> get_string
-          |> fun address -> return {address}
+          from_string body |> get_dict |> List.assoc "address" |> get_string
+          |> fun address -> return { address }
```
Comment (knuton) [inline controller/bindings/zerotier/zerotier.ml:23]: This seems quite unfortunate to me, pipelines are sometimes on multiple lines, sometimes not, and overall harder for me to read.

Have you considered [`break-infix = wrap-or-vertical`](https://ocaml.org/p/ocamlformat/latest/doc/manpage_ocamlformat.html#:~:text=value%20is%20wrap.-,%2D%2Dbreak%2Dinfix%3D,-%7Bwrap%7Cfit%2Dor)?

File: controller/bindings/curl/curl.ml [hunk 7]
```diff
@@ -60,72 +50,63 @@ let exec cmd =
...
-      ; (headers
-          |> List.map (fun (k, v) -> [| "--header"; (k ^ ":" ^ v) |])
-          |> Array.concat)
-      ; (match data with
-        | Some d -> [| "--data"; d |]
-        | None -> [| |])
-      ; Base.List.to_array options
-      ])
+    ( ""
+    , (* path to Curl executable (uses PATH if empty string) *)
```
Comment (knuton) [inline controller/bindings/curl/curl.ml:66]: Should probably move this before the comma so it stays close to the string literal it refers to.

File: controller/bindings/connman/connman.mli (hunk 5 above)
Comment (yfyf) [reply controller/bindings/connman/connman.mli:48]: Putting it above actually yields a warning: `Warning 50 [unexpected-docstring]: unattached documentation comment `

File: controller/bindings/curl/curl.ml (hunk 7 above)
Comment (yfyf) [reply controller/bindings/curl/curl.ml:66]: Yes, will do some manual fixes once we settle on the final formatting rules.

**Comment by yfyf** [comment]
>  We’ll need a CI check to enforce the formatting afterward.

Added a simple check in d672360

> And also consider formating in dev-server.

My personal preference would be not to have anything interrupting development due to incorrect formatting, so I would vote for not adding it there.

**Comment by knuton** [comment]
> My personal preference would be not to have anything interrupting development due to incorrect formatting, so I would vote for not adding it there.

What we normally have is not a check that fails, but an option to auto-format as part of the dev loop `AUTO_FORMAT=1 bin/dev-server`. Some in the team use editor integration, some let the dev server do it.

**Comment by yfyf** [comment]
> What we normally have is not a check that fails, but an option to auto-format as part of the dev loop AUTO_FORMAT=1 bin/dev-server. Some in the team use editor integration, some let the dev server do it.

Added in 3c8004e. It will cause a double reload/restart (due to the written changes), but I guess that's OK?

If this was the last thing, I can squash the commits, deal with `.git-blame-ignore-revs` and manicure the comment / multi-line string formatting.

**Comment by knuton** [comment]
> If this was the last thing, I can squash the commits, deal with `.git-blame-ignore-revs` and manicure the comment / multi-line string formatting.

I think so, but do we need to get https://github.com/dividat/playos/pull/209 through first? Or how would we resolve conflicts?

**Comment by yfyf** [comment]
> I think so, but do we need to get https://github.com/dividat/playos/pull/209 through first?

Yes, that would certainly be preferable!

**Comment by knuton** [comment]
https://github.com/dividat/playos/pull/209 is merged

**Comment by yfyf** [comment]
squashed and rebased on top of main, ready for merging.

**Review by knuton — APPROVED** [review]

**Pull request merged**